File: app/src/main/java/com/scop/org/minesweeper/generators/RandomCheckedGenerator.java
package com.scop.org.minesweeper.generators;

import com.scop.org.minesweeper.control.MainLogic;
import com.scop.org.minesweeper.control.Settings;
import com.scop.org.minesweeper.elements.Grid;
import com.scop.org.minesweeper.elements.Tile;
import com.scop.org.minesweeper.generators.solver.BasicSolver;
import com.scop.org.minesweeper.generators.solver.DeepSolver;
import com.scop.org.minesweeper.generators.solver.PossibleSolver;
import com.scop.org.minesweeper.generators.solver.Sketch;
import com.scop.org.minesweeper.generators.solver.Solver;
import com.scop.org.minesweeper.utils.GridUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class RandomCheckedGenerator extends RandomGenerator {
	private int selectedSafeTile;


	@Override
	public void generateNewGrid(Grid grid, int bombs, FinishCallback cb) {
		new Thread(){
			@Override
			public void run(){
				generateNewRandomGrid(grid, bombs);
				selectedSafeTile = GridUtils.findSafeOpenTileIdx(grid);

				Grid originalGrid = grid;
				int numBombsLeft = 1;

				while (numBombsLeft != 0) {
					Set<Integer> bombsLeft = solve(originalGrid.clone());
					numBombsLeft = bombsLeft == null? Integer.MAX_VALUE : bombsLeft.size();

					if (numBombsLeft > 0) {
						if (numBombsLeft < bombs/20) {
							swapBombs(grid, bombsLeft);
							originalGrid = grid;

						} else {
							generateNewRandomGrid(grid, bombs);
							originalGrid = grid;
							selectedSafeTile = GridUtils.findSafeOpenTileIdx(grid);
						}

					}
				}

				if (Settings.getInstance().isFirstOpen()) {
					MainLogic ml = new MainLogic(grid);
					Tile t = grid.getGrid().get(selectedSafeTile);
					ml.reveal(t);
				}

				cb.finished();
			}
		}.start();
	}


	public void swapBombs(Grid grid, Set<Integer> toReplace) {
		List<Integer> protectedTiles = GridUtils.getNeighborsIdx(grid, grid.getGrid().get(selectedSafeTile));
		protectedTiles.add(selectedSafeTile);

		for (int rplc : toReplace) {
			int aim;
			do {
				aim = GridUtils.findSafeOpenTileIdx(grid);
			} while (protectedTiles.indexOf(aim) >= 0);

			Tile t0 = grid.getGrid().get(rplc);
			t0.plantBomb(false);
			GridUtils.getNeighbors(grid, t0).forEach(Tile::doesntHaveBombNear);

			Tile tf = grid.getGrid().get(aim);
			tf.plantBomb();
			GridUtils.getNeighbors(grid, tf).forEach(Tile::hasBombNear);
		}
	}

	public Set<Integer> solve(Grid grid){
		MainLogic ml = new MainLogic(grid);

		Tile t = grid.getGrid().get(selectedSafeTile);
		ml.reveal(t);

		Sketch sketch = new Sketch(grid);

		Solver.setSolver(ml,sketch);

		List<Solver> solvers = new ArrayList<>();
		solvers.add(new BasicSolver());
		solvers.add(new PossibleSolver());
		solvers.add(new DeepSolver());

		while(
			solvers.stream().anyMatch(Solver::analyze)
		);
		return Solver.getLeftBombs();
	}
}

File: app/src/main/java/org/oar/minesweeper/generators/RandomCheckedGenerator.kt
package org.oar.minesweeper.generators

import org.oar.minesweeper.grid.GameLogic
import org.oar.minesweeper.models.Grid
import org.oar.minesweeper.models.GridGenerationDetails
import org.oar.minesweeper.models.Tile
import org.oar.minesweeper.generators.solver.*
import org.oar.minesweeper.utils.GridUtils.findSafeOpenTileIdx
import org.oar.minesweeper.utils.GridUtils.getNeighbors
import org.oar.minesweeper.utils.GridUtils.getNeighborsIdx

open class RandomCheckedGenerator : RandomGenerator() {

    companion object {
        private const val MAX_SWAPPING_RETRY = 5
    }

    protected var selectedSafeTile = 0
    private var sketch: Sketch? = null
    override fun generateNewGrid(grid: Grid, onFinish: (GridGenerationDetails) -> Unit) {
        Thread {
            generateNewRandomGrid(grid)
            selectedSafeTile = grid.findSafeOpenTileIdx()

            var originalGrid = grid
            var numBombsLeft = 1
            var retrySwapping = MAX_SWAPPING_RETRY

            while (numBombsLeft != 0) {
                val bombsLeft = solve(originalGrid.clone())
                numBombsLeft = bombsLeft?.size ?: Int.MAX_VALUE

                if (numBombsLeft > 0) {
                    if (numBombsLeft < grid.bombs / 20 && retrySwapping > 0) {
                        retrySwapping--
                        swapBombs(grid, bombsLeft!!)
                        originalGrid = grid

                    } else {
                        generateNewRandomGrid(grid)
                        originalGrid = grid
                        selectedSafeTile = grid.findSafeOpenTileIdx()
                        retrySwapping = MAX_SWAPPING_RETRY
                    }
                }
            }

            onFinish(
                GridGenerationDetails(selectedSafeTile)
            )
        }.start()
    }

    override fun forceCleanSpot(grid: Grid, tile: Tile) {
        throw RuntimeException("Function not supported for solvable grids")
    }

    fun swapBombs(grid: Grid, toReplace: Set<Int>) {
        val protectedTiles = grid.getNeighborsIdx(grid.tiles[selectedSafeTile])
        protectedTiles.add(selectedSafeTile)

        for (rplc in toReplace) {
            var aim: Int
            do {
                aim = grid.findSafeOpenTileIdx()
            } while (protectedTiles.indexOf(aim) >= 0)

            val t0: Tile = grid.tiles[rplc]
            t0.hasBomb = false
            grid.getNeighbors(t0).forEach { it.bombsNear-- }

            val tf: Tile = grid.tiles[aim]
            tf.hasBomb = true
            grid.getNeighbors(tf).forEach { it.bombsNear++ }
        }
    }

    fun solve(grid: Grid): Set<Int>? {
        val ml = GameLogic(grid)

        val t: Tile = grid.tiles[selectedSafeTile]
        ml.reveal(t)

        sketch = Sketch(grid)
        Solver.setSolver(ml, sketch!!)

        val solvers: MutableList<Solver> = mutableListOf(
            BasicSolver(),
            PossibleSolver(),
            DeepSolver(),
            AdvancedSolver(),
        )

        @Suppress("ControlFlowWithEmptyBody")
        while (solvers.any { solver -> solver.analyze() });

        return Solver.leftBombs
    }
}